src/combinator.rs

```diff
@@ -1618,6 +1615,61 @@ where
     go_extra!(C);
 }
 
+/// See [`IterParser::collect_exactly`]
+pub struct CollectExactly<A, O, C, const N: usize> {
+    pub(crate) parser: A,
+    pub(crate) phantom: PhantomData<(O, C)>,
+}
+
+impl<A: Copy, O, C, const N: usize> Copy for CollectExactly<A, O, C, N> {}
+impl<A: Clone, O, C, const N: usize> Clone for CollectExactly<A, O, C, N> {
+    fn clone(&self) -> Self {
+        Self {
+            parser: self.parser.clone(),
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<'a, I, O, E, A, C, const N: usize> Parser<'a, I, C, E> for CollectExactly<A, O, C, N>
+where
+    I: Input<'a>,
+    E: ParserExtra<'a, I>,
+    A: IterParser<'a, I, O, E>,
+    C: ContainerExactly<O, N>,
+{
+    #[inline]
+    fn go<M: Mode>(&self, inp: &mut InputRef<'a, '_, I, E>) -> PResult<M, C> {
+        let before = inp.offset();
+        let mut output = M::bind(|| C::uninit());
+        let mut iter_state = self.parser.make_iter::<M>(inp)?;
+        for idx in 0..N {
+            match self.parser.next::<M>(inp, &mut iter_state) {
+                Ok(Some(out)) => {
+                    M::combine_mut(&mut output, out, |c, out| C::write(c, idx, out));
+                }
+                Ok(None) => {
+                    inp.add_alt(
```

Comment:
Would be nice to only do this if there's no error? I don't know

Reply:
Another of these "we need to generate an error but we don't know how" cases :(